### PR TITLE
[BACK-77] Integrate twitter, categories and fork history in dbUpdater

### PR DESCRIPTION
**Pull request by thomas-woehrle**
### Description of the issue
Integrate twitter data, the forkHistory and categories, so that the frontend can access that.
The new contributor count should be stored instead of one.
Store relative metrics on the db as well. 
Remove organizations and associated persons that do not own projects on the db anymore.

### How I implemented
- integrate twitter data, forkHistory and categories through functions in `updateProject.ts`
- integrate the new contributorCount in `updateProjectGithubStats()`
- integrate calculation of relative metrics in `updateProjectGithubStats()`
- adjust supabase schema 
- implement functions in `supabaseUtils.ts` that find and remove stale associatedPersons and stale organizations

### Other
- Forkhistory uses mock data for now, but only one line needs to be changed for that to work
- the biggest part of the LOCs changed is the autogenerated supabase schema

### CODE REVIEW

File: packages/graphql_server/src/updateProject.ts [hunk 1]
```diff
@@ -144,6 +144,54 @@ const updateProjectFounders = async (repoName: string, owner: string) => {
...
+  }
+
+  // *0.9 because the contributor count is not 100% accurate. It is an approximation.
+  // *10 /10 to round to one decimal place
+  //@Todo: refine approximation approach
+  const contributorCount =
+    Math.round(await getContributorCount(owner, name, process.env.GITHUB_API_TOKEN)) * 0.9
+
+  // *10, round, and /10 to round to one decimal place
+  const issuesPerContributor =
```
Comment (pczern) [inline packages/graphql_server/src/updateProject.ts:178]: what exactly will *10 do; if it is 4.8 isn't that an accurate number important to investors?

File: packages/graphql_server/src/updateProject.ts [hunk 2]
```diff
@@ -144,6 +144,54 @@ const updateProjectFounders = async (repoName: string, owner: string) => {
...
+ * Updates the github stats of a repo.
+ * Not the star-history though.
+ * @param {string} repoName - The name of the repo.
+ * @param {string} owner - The name of the owner of the repo.
+ */
+const updateProjectGithubStats = async (name: string, owner: string) => {
+  if (!(await repoIsAlreadyInDB(name, owner))) return
+  const githubStats: GitHubInfo | null = await getGithubData(name, owner)
+  if (!githubStats) {
+    console.log('Could not get github stats for ', name, 'owned by', owner)
```
Comment (pczern) [inline packages/graphql_server/src/updateProject.ts:167]: now way to access error from outside of method

File: packages/graphql_server/src/updateProject.ts [hunk 3]
```diff
@@ -144,6 +144,54 @@ const updateProjectFounders = async (repoName: string, owner: string) => {
   }
 }
 
+const updateProjectForkHistory = async (repoName: string, owner: string) => {
+  if (!(await repoIsAlreadyInDB(repoName, owner))) {
+    return
+  }
+  //@Todo: get real fork history -> waiting for maxis PR
```
Comment (pczern) [inline packages/graphql_server/src/updateProject.ts:151]: // `@TODO`

File: packages/graphql_server/src/supabaseUtils.ts [hunk 4]
```diff
@@ -151,6 +185,37 @@ const formatLinkedInCompanyData = (linkedInData: LinkedInCompanyProfile): Organi
   }
 }
 
+const getListOfProjectsFoundedByUser = async (githubUsername: string) => {
+  const personID = await getPersonID(githubUsername)
+
+  const { data: projects } = await supabase
```
Comment (pczern) [inline packages/graphql_server/src/supabaseUtils.ts:191]: where is the error thrown?

File: packages/graphql_server/src/api/openAIApi.ts [hunk 5]
```diff
@@ -141,20 +141,20 @@ export const getCategoryFromGPT = async (topics: string[] | null, description: s
...
+      return Category[Number(categoryNumber)]
     })
   } catch (error) {
-    return Topic['9']
+    return [Category['9']]
   }
 }
 
 // iterating through enums needs to be done like that
 const createCategorizationPrompt = () => {
```
Comment (pczern) [inline packages/graphql_server/src/api/openAIApi.ts:152]: this method should be rewritten completely, I think the approach is way to difficult

File: packages/graphql_server/src/updateProject.ts [hunk 6]
```diff
@@ -44,17 +49,33 @@ const updateAllProjectInfo = async (
...
     return
   }
+  await updateProjectGithubStats(repoName, owner)
   await updateProjectELI5(repoName, owner)
   await updateProjectFounders(repoName, owner)
-  await updateProjectGithubStats(repoName, owner)
   await updateProjectLinkedInData(owner)
   await updateProjectSentiment(repoName, owner)
   await updateProjectStarHistory(repoName, owner)
+  await updateProjectForkHistory(repoName, owner)
```
Comment (pczern) [inline packages/graphql_server/src/updateProject.ts:58]: this is wrong

File: packages/graphql_server/src/updateProject.ts (hunk 1 above)
Comment (thomas-woehrle) [reply packages/graphql_server/src/updateProject.ts:178]: Math.round(x *10) / 10 is my way of rounding x to one decimal point. 
for example if x is 4.84 this will return 4.8

File: packages/graphql_server/src/updateProject.ts [hunk 7]
```diff
@@ -144,6 +144,54 @@ const updateProjectFounders = async (repoName: string, owner: string) => {
...
+    Math.round((githubStats.issues.totalCount / contributorCount) * 10) / 10
+  const forksPerContributor = Math.round((githubStats.forkCount / contributorCount) * 10) / 10
+
+  const updated = await updateSupabaseProject(name, owner, {
+    ...formatGithubStats(githubStats),
+    issues_per_contributor: issuesPerContributor,
+    forks_per_contributor: forksPerContributor
+  })
+
+  if (!updated) {
```
[the remark this answers is not in the document]
Comment (thomas-woehrle) [reply packages/graphql_server/src/updateProject.ts:188]: you can't. 
it is bad design as I realize now but won't change before the midterm review

File: packages/graphql_server/src/updateProject.ts (hunk 2 above)
Comment (thomas-woehrle) [reply packages/graphql_server/src/updateProject.ts:167]: it is bad design as I realize now but won't change before the midterm review

File: packages/graphql_server/src/updateProject.ts (hunk 6 above)
Comment (thomas-woehrle) [reply packages/graphql_server/src/updateProject.ts:58]: There are design flaws absolutely. But what we do here is updating the info that we get. If let's say we don't get the hackernews sentiment of a project we would still want to add the project to the db with everything else

File: packages/graphql_server/src/utils.ts [hunk 8]
```diff
@@ -79,3 +79,102 @@ export const parseGitHubUrl = (url: string) => {
 
   return null
 }
+
+export const mockForkHistory = [
```
[the remark this answers is not in the document]
Comment (thomas-woehrle) [reply packages/graphql_server/src/utils.ts]: correct place would be something like /assets, but it wont be there for long so lets leave that out for now

**Comment by thomas-woehrle** [comment]
How is it possible that this fails, when I can run it locally without failure and the linting gives no errors locally as well?
Screenshot 2023-06-22 at 18 31 56

**Review by acreturs — APPROVED** [review]

**Review by 2015builder — APPROVED** [review]
